extension/source/Controllers/Network/NetworkController.ts

```diff
@@ -0,0 +1,249 @@
+   */
+  private ethQuery: EthQuery;
+
+  private _baseProviderHandlers: IProviderHandlers;
+
+  constructor({
+    config,
+    state,
+  }: {
+    config?: Partial<NetworkConfig>;
+    state?: Partial<NetworkState>;
+  }) {
+    super({ config, state });
+    this.defaultState = {
+      chainId: 'loading',
+      properties: {
+        EIPS: { 1559: undefined },
+      },
+      providerConfig: SUPPORTED_NETWORKS[CHAINS.MAINNET],
+    };
+    this.initialize();
+    // when a new network is set, we set to loading first and then when connection succeeds, we update the network
+  }
+
+  getNetworkIdentifier(): string {
+    return this.state.chainId;
+  }
+
+  /**
+   * Called by orchestrator once while initializing the class
+   * @param providerHandlers - JRPC handlers for provider
+   * @returns - provider - Returns the providerProxy
+   */
+  public initializeProvider(
+    providerHandlers: IProviderHandlers,
+  ): SafeEventEmitterProvider {
+    this._baseProviderHandlers = providerHandlers;
+    this.configureProvider();
+    this.lookupNetwork(); // Not awaiting this, because we don't want to block the initialization
+    return this._providerProxy;
+  }
+
+  getProvider(): SafeEventEmitterProvider {
+    return this._providerProxy;
+  }
+
+  setProviderConfig(config: ProviderConfig): void {
+    this.update({
+      providerConfig: { ...config },
+    });
+    this.refreshNetwork();
+  }
+
+  getProviderConfig(): ProviderConfig {
+    return this.state.providerConfig;
+  }
+
+  /**
+   * Refreshes the current network code
+   */
+  async lookupNetwork(): Promise<void> {
+    const { rpcTarget, chainId } = this.getProviderConfig();
+    if (!chainId || !rpcTarget || !this._provider) {
+      this.update({ chainId: 'loading', properties: {} });
+      return;
+    }
+    const query = this.ethQuery;
+    if (!query) {
+      return;
+    }
+    const releaseLock = await this.mutex.acquire();
+    return new Promise((resolve, reject) => {
+      // info_get_status
+      query.sendAsync(
+        { method: 'net_version' },
+        (error: Error, network: unknown) => {
+          releaseLock();
+          if (error) {
+            this.update({
+              chainId: 'loading',
+              properties: {
+                EIPS: { 1559: undefined },
+              },
+            });
+            reject(error);
+          }
+
+          this.update({
+            // Network is returned as a string (base 10)
+            chainId: `0x${Number.parseInt(network as string, 16).toString()}`,
+          });
+          // Don't need to wait for this
+          this.getEIP1559Compatibility();
+          this.emit('networkDidChange');
+          resolve();
+        },
+      );
+    });
+  }
+
+  async getEIP1559Compatibility(): Promise<boolean> {
+    const { EIPS } = this.state.properties;
+    // log.info('checking eip 1559 compatibility', EIPS[1559])
+    if ((EIPS as any)[1559] !== undefined) {
+      return (EIPS as any)[1559];
+    }
```

Comment:
```suggestion
    const { EIPS } = this.state.properties as any;
    // log.info('checking eip 1559 compatibility', EIPS[1559])
    if (EIPS[1559] !== undefined) {
      return Boolean(EIPS[1559]);
    }
```

Reply:
If `Record<string, boolean | undefined>` is an accurate type for `EIPS` then it would be nice to add that typing to `this.state.properties` rather than do casts and conversions here.

Reply:
addressed by adding full types